packages/styleguide/src/components/List/List.tsx

```diff
@@ -142,12 +142,22 @@ export const ListItem = ({
 type ListProps = {
   children: React.ReactNode
   data: {
-    ordered?: boolean
-    start?: number
     compact?: boolean
+    ordered?: boolean
   }
-  attributes?: React.HTMLAttributes<HTMLUListElement | HTMLOListElement>
-}
+} & (
+  | {
+      data: {
+        ordered: true
+        start?: number
+      }
+      attributes?: React.ComponentProps<typeof OrderedList>['attributes']
+    }
+  | {
+      data: { ordered?: false }
+      attributes?: React.ComponentProps<typeof UnorderedList>['attributes']
+    }
+)
 
 export const List = ({ children, data, attributes = {} }: ListProps) =>
```

Comment:
@jstcki This causes an interesting issue with TypeScript, could you take a look?

```tsx
@jstcki Interesting TS Issue:

```tsx
I ran into an interesting issue with discriminary unions in a react component .
This is the component:
```tsx
type ListProps = {
  children: React.ReactNode
  data: {
    compact?: boolean
    ordered?: boolean
  }
} & (
  | {
      data: {
        ordered: true
        start?: number
      }
      attributes?: React.ComponentProps<typeof OrderedList>['attributes']
    }
  | {
      data: { ordered?: false }
      attributes?: React.ComponentProps<typeof UnorderedList>['attributes']
    }
)

export const List = ({ children, data, attributes = {} }: ListProps) =>
  data.ordered ? (
    <OrderedList
      start={data.start}
      compact={data.compact}
      attributes={attributes}
    >
      {children}
    </OrderedList>
  ) : (
    <UnorderedList compact={data.compact} attributes={attributes}>
      {children}
    </UnorderedList>
  )
```
However this now yields the following issue: 
```tsx
const ListAbstraction = ({…, bool?: boolean}) => <List data={{ ordered: bool }}>{…}</List> // Type 'boolean' is not assignable to type 'false'.
```
Any Idea how this could be further improved?
I would've assumed for this work work as expected, as `{ 'ordered: true' } | { 'ordered?: false' }` in my opinion would satisfy `{ ordered?: boolean }`.
Seems this bring TS to it's limits ^^ (perhaps I am missing something though?)
```

Reply:
Oh no… Works fine in a [sandbox environment](https://stackblitz.com/edit/stackblitz-starters-ltlz6y?file=src%2FList.tsx)… our setup seems to be causing issues once again…

Reply:
Creating a proper tagged union would work (but that would need changes elsewhere, so probably not desirable):

```tsx
type ListProps =
  | {
      listType: 'ordered'
      children: ReactNode
      data: {
        ordered?: boolean
        start?: number
        compact?: boolean
      }
      attributes?: React.ComponentProps<typeof OrderedList>['attributes']
    }
  | {
      listType: 'unordered'
      children: ReactNode
      data: { ordered?: boolean; compact?: boolean }
      attributes?: React.ComponentProps<typeof UnorderedList>['attributes']
    }

export const List = ({
  children,
  listType,
  data,
  attributes = {},
}: ListProps) =>
  listType === 'ordered' ? (
    <OrderedList
      start={data.start}
      compact={data.compact}
      attributes={attributes}
    >
      {children}
    </OrderedList>
  ) : (
    <UnorderedList compact={data.compact} attributes={attributes}>
      {children}
    </UnorderedList>
  )
```

Reply:
So, it appears the problem is when you pass an explicitly typed `boolean` to literal `true` or `false`. See [this example](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAJQKYEMDGMA0cDez0wByEAJkgL5wBmUEIcARFKhgwFBswCeYScA8gBlgAZxgAFWmBFwAvLjZw4aABbAANiWYA7AFyIWRUkkVwSKGCn05TS6GWYl9MKAFcTSpWJSwA-Pu1XEAAjJChbZTowAn84YIgIdVRtU3I2NM4ePgBVYTFJCGk5BSVVDS0kPQMCYjJTc0trOHswpBJYqhR1ESQAbkjwGP14xOS4NIzuXjg8iSkZeSFROcKZAB84XOWC6Q40CG0xGeXigAowed1ZnZEASjkAPjgAHgecC9WAOgaUT5bHOC+Rj-NoMfQMVzaEEkBjkZ4AegeewOR34AGsUFxZmccNDyLpcVAHG1-CMkihtOR7rInq8Is9sT9ZDhCcSnC53ORyA9nupgA8VEguAi+Tz4bMkZ4XoyLChmazWk5Ot0KNzefzBcL4aKERK2AikWx4fDIod4MQYOjMdj5KcFY58faSboyckqY9pScmSy8WrRZqRfzdctDfszXAAILdCBWrEnW1Okj43DQ2KuinumnNImKwGeo7exPONyqnn+oWBsUSuD6Ble2Xy6H6ZU9LlljUV7VB8UhjhAA).

Doing this at the call site fixes it:

```tsx
export const List: React.FC<{
  children?: ReactNode
  ordered?: boolean
  [x: string]: unknown
}> = ({ children, ordered, attributes = {}, ...props }) => {
  return ordered ? (
    <InnerList {...props} data={{ ordered: true }} attributes={attributes}>
      {children}
    </InnerList>
  ) : (
    <InnerList {...props} data={{ ordered: false }} attributes={attributes}>
      {children}
    </InnerList>
  )
}
```

Reply:
> So, it appears the problem is when you pass an explicitly typed `boolean` to literal `true` or `false`. See [this example](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAJQKYEMDGMA0cDez0wByEAJkgL5wBmUEIcARFKhgwFBswCeYScA8gBlgAZxgAFWmBFwAvLjZw4aABbAANiWYA7AFyIWRUkkVwSKGCn05TS6GWYl9MKAFcTSpWJSwA-Pu1XEAAjJChbZTowAn84YIgIdVRtU3I2NM4ePgBVYTFJCGk5BSVVDS0kPQMCYjJTc0trOHswpBJYqhR1ESQAbkjwGP14xOS4NIzuXjg8iSkZeSFROcKZAB84XOWC6Q40CG0xGeXigAowed1ZnZEASjkAPjgAHgecC9WAOgaUT5bHOC+Rj-NoMfQMVzaEEkBjkZ4AegeewOR34AGsUFxZmccNDyLpcVAHG1-CMkihtOR7rInq8Is9sT9ZDhCcSnC53ORyA9nupgA8VEguAi+Tz4bMkZ4XoyLChmazWk5Ot0KNzefzBcL4aKERK2AikWx4fDIod4MQYOjMdj5KcFY58faSboyckqY9pScmSy8WrRZqRfzdctDfszXAAILdCBWrEnW1Okj43DQ2KuinumnNImKwGeo7exPONyqnn+oWBsUSuD6Ble2Xy6H6ZU9LlljUV7VB8UhjhAA).
> 
> Doing this at the call site fixes it:
> 
> ```tsx
> export const List: React.FC<{
>   children?: ReactNode
>   ordered?: boolean
>   [x: string]: unknown
> }> = ({ children, ordered, attributes = {}, ...props }) => {
>   return ordered ? (
>     <InnerList {...props} data={{ ordered: true }} attributes={attributes}>
>       {children}
>     </InnerList>
>   ) : (
>     <InnerList {...props} data={{ ordered: false }} attributes={attributes}>
>       {children}
>     </InnerList>
>   )
> }
> ```

What I don't understand, is why it works fine in sandbox environments 🤔

Reply:
> What I don't understand, is why it works fine in sandbox environments 🤔

It doesn't! It all works fine if you pass an explicit true or false to `<List>` but what doesn't work is when it's wrapped in another component which passes on a `boolean` prop.

![image](https://github.com/republik/plattform/assets/70969/00d39748-7b77-4321-b6a0-2f9d77e104b5)